projects/github-actions/repo-gardening/src/tasks/add-labels/index.js

```diff
@@ -336,20 +337,45 @@ async function addLabels( payload, octokit ) {
 	// If the PR title includes the word "revert", mark it as such.
 	const isRevert = title.toLowerCase().includes( 'revert' );
 
-	const labels = await getLabelsToAdd( octokit, owner.login, name, number, isDraft, isRevert );
+	const currentLabels = payload.pull_request.labels.map( l => l.name );
+	let labelsToAdd = await getLabelsToAdd(
+		octokit,
+		owner.login,
+		name,
+		number,
+		isDraft,
+		isRevert,
+		currentLabels
+	);
 
-	if ( ! labels.length ) {
-		debug( 'add-labels: Could not find labels to add to that PR. Aborting' );
+	// Nothing new was added, so abort.
+	if ( labelsToAdd.length === currentLabels.length ) {
+		debug( 'add-labels: No new labels to add to that PR. Aborting.' );
 		return;
 	}
 
-	debug( `add-labels: Adding labels ${ labels } to PR #${ number }` );
+	// GitHub allows 100 labels on a PR. Limit to less than that to allow for additional labels elsewhere.
+	const maxLabels = 90;
+	const bigLabel = 'All the things (>' + maxLabels + ' labels)';
+	if ( labelsToAdd.length > maxLabels ) {
+		debug(
+			'add-labels: GitHub only allows 100 labels on a PR, so limiting to the first ' +
+				maxLabels +
+				'.'
+		);
+		labelsToAdd.splice( maxLabels );
+		labelsToAdd.push( bigLabel );
+	} else if ( labelsToAdd.includes( bigLabel ) ) {
+		labelsToAdd = labelsToAdd.filter( label => label !== bigLabel );
+	}
+
+	debug( `add-labels: Adding labels ${ labelsToAdd } to PR #${ number }` );
 
-	await octokit.rest.issues.addLabels( {
+	await octokit.rest.issues.setLabels( {
```

Comment:
> this might stomp on those manual additions

Yeah, that's true. I was hoping to avoid an extra query, but it'd be safer to query at runtime (also to prevent stomping on manual label removals).

> It might be safer to remove the labels we want to remove

I assume by this you mean get a diff between the (real) current labels and the ones we want to add, and then add just those we want to add?

Reply:
Pretty much, yeah. There's also the removing of the `bigLabel` label to happen somewhere.

Reply:
9661caa15c